src/lib/data/alignment-group.js

```diff
@@ -122,36 +157,179 @@ export default class AlignmentGroup {
+   * @returns { Object }
+   *          { AlignmentGroup } tokensGroup - group that was merged
+   *          { Number } indexDeleted - place in group list
+   */
+  unmerge (step) {
+    const tokensGroup = step.type === 'merge' ? step.token : []
+
+    for (let i = 0; i < tokensGroup.origin.length; i++) {
+      const tokenIdWord = tokensGroup.origin[i]
+      const tokenIndex = this.origin.findIndex(tokenId => tokenId === tokenIdWord)
+      if (tokenIndex >= 0) {
+        this.origin.splice(tokenIndex, 1)
+      }
+    }
+
+    for (let i = 0; i < tokensGroup.target.length; i++) {
+      const tokenIdWord = tokensGroup.target[i]
+      const tokenIndex = this.target.findIndex(tokenId => tokenId === tokenIdWord)
+      if (tokenIndex >= 0) {
+        this.target.splice(tokenIndex, 1)
+      }
+    }
+
+    this.defineFirstStepToken()
+
+    return {
+      tokensGroup,
+      indexDeleted: step.indexDeleted
+    }
+  }
+
+  /**
+   * Finds token with the given idWord using steps
+   * @param {String} idWord
+   * @retuns { Token|null }
+   */
+  findTokenByIdWord (idWord) {
+    const step = this.steps.find(step => step.idWord === idWord)
+    return step ? step.token : null
   }
+
+  /**
+   * Step back
+   * @retuns { Array(Object) } - results of undone steps, for example result of unmerge action
+   */
+  undo () {
+    if (this.steps.length > 1 && this.currentStepIndex > 0) {
+      const results = this.alignToStep(this.currentStepIndex - 1)
+      return results
+    } else {
+      console.error(L10nSingleton.getMsgS('ALIGNMENT_GROUP_UNDO_ERROR'))
+    }
+  }
+
+  /**
+   * Step forward
+   */
+  redo () {
+    if (this.currentStepIndex < (this.steps.length - 1)) {
+      this.alignToStep(this.currentStepIndex + 1)
+    } else {
+      console.error(L10nSingleton.getMsgS('ALIGNMENT_GROUP_REDO_ERROR'))
+    }
+  }
+
+  /**
+   * Defines current position in the step tracker and apply/remove step actions according to the position
+   * @param {Number} stepIndex
+   * @retuns { Array(Object) } - results of undone steps, for example result of unmerge action
+   */
+  alignToStep (stepIndex) {
+    let results = [] // eslint-disable-line prefer-const
+    if (this.currentStepIndex > stepIndex) {
+      for (let i = this.currentStepIndex; i > stepIndex; i--) {
+        const result = this.removeStepAction(i)
+        if (result && result.tokensGroup) { results.push(result) }
+      }
+    } else if (this.currentStepIndex < stepIndex) {
+      for (let i = this.currentStepIndex + 1; i <= stepIndex; i++) {
+        const result = this.applyStepAction(i)
+        if (result && result.indexDeleted) { results.push(result) }
+      }
+    }
+    this.currentStepIndex = stepIndex
+    return results
+  }
+
+  /**
+   * Removes the step action,
+   * the following actions are defined - add, remove, merge
+   * @param {Number} stepIndex
+   * @retuns {Object} - results of undone steps, for now it could return only the result of unmerge action

```

Comment:
I didn't fully understand your thought here.

At the previous iteration I coded that each action (not only merge) returned something.
But this way we received really different results:
- for `add` - an array of deleted items 
- for `remove` - a number = a new length of the array
- for `merge` - an object { tokensGroup, indexDeleted }

We would use only the last case in the next code's steps.
As for `applyStepAction` - we won't return anything to be used in the code.

And that's why I decided to remove unuseful returns to reduce complexity and unuseful code statements.

@balmas , what exactly do you suggest me to change here? to use return in each statement, describe all variants of return and notice that they won't be used in next code's steps?

Reply:
I'm not sure. I think it goes along with my suggestion that we might want to make the history steps first class objects, and maybe also a separate object type for a step removal.    Or we could just make it something simple like an object with a status and data e.g.
```
{ status: boolean, //(true if removal succeeded otherwise false)
  data: {} // (only populated if there is data to return
}
```
I realize that it's effectively the same thing as you have but it keeps the method signature consistent and I think is a little more robust

Reply:
ok, let's add such a simple object for now, if you think that it would be valuable